src/transformers/models/t5/modeling_t5.py

```diff
@@ -238,8 +238,11 @@ def __init__(self, hidden_size, eps=1e-6):
         self.variance_epsilon = eps
 
     def forward(self, hidden_states):
-        # layer norm should always be calculated in float32
-        variance = hidden_states.to(torch.float32).pow(2).mean(-1, keepdim=True)
+        # layer norm should be calculated in float32 except for bfloat16
+        if hidden_states.dtype == torch.bfloat16:
+            variance = hidden_states.pow(2).mean(-1, keepdim=True)
+        else:
+            variance = hidden_states.to(torch.float32).pow(2).mean(-1, keepdim=True)
```

Comment:
Could you please point me to why do you propose that it is safe to accumulate in bf16? You do have a higher numerical range but the precision is much worse than fp16.

Let's perhaps run a practical test and compare the error between doing the above in fp32 and bf16?

Reply:
Update: I checked with NVIDIA and they confirmed that this won't work. This operation has to happen in fp32.

pytorch's layernorm has this issue as well, and it has been fixed only in the last month, so any current pytorch version will do the wrong thing with fp16 or bf16 inputs.

This is where it has been fixed: https://github.com/pytorch/pytorch/pull/66920 - I'd imagine we will see it in pt-1.10.1

Reply:
Thank you for the quick review and for catching that issue. I've reverted the variance computation to the initial `float32`.